### PR TITLE
Fix reservation cloning

### PR DESCRIPTION
**Pull request by nop33**
Make necessary start and end datetime adjustments when cloning a reservation.

Closes SNow ticket `INC0962765`

### CODE REVIEW

File: indico/modules/rb/controllers/user/reservations.py [hunk 1]
```diff
@@ -461,6 +461,19 @@ def _make_form(self):
         self.past_date = self.date_changed = False
         changes = {'room_id': self._room.id}
 
+        # Move the start date to the current day, adjust the end date
+        # accordingly and if the start datetime is still in the past, change the
+        # start and end time to the current (rounded up) time.
+        reservation_duration = self._reservation.end_dt - self._reservation.start_dt
+        date_delta = datetime.now().date() - self._reservation.start_dt.date()
```
Comment (ThiefMaster) [inline indico/modules/rb/controllers/user/reservations.py]: `date.today()`

File: indico/modules/rb/controllers/user/reservations.py [hunk 2]
```diff
@@ -461,6 +461,19 @@ def _make_form(self):
...
+        changes['start_dt'] = self._reservation.start_dt + date_delta
+        changes['end_dt'] = changes['start_dt'] + reservation_duration
+        if changes['start_dt'] < datetime.now():
+            date_now = round_up_to_minutes(datetime.now(), 15) + timedelta(minutes=15)
```
Comment (ThiefMaster) [inline indico/modules/rb/controllers/user/reservations.py]: wht not `start_dt`?

Comment (nop33) [inline indico/modules/rb/controllers/user/reservations.py]: why not you going to rest for a day? :mask:

File: indico/modules/rb/controllers/user/reservations.py [hunk 3]
```diff
@@ -461,6 +461,19 @@ def _make_form(self):
         self.past_date = self.date_changed = False
         changes = {'room_id': self._room.id}
 
+        # Move the start date to the current day, adjust the end date
```
Comment (OmeGak) [inline indico/modules/rb/controllers/user/reservations.py]: If we need this massive comment I would move the logic to a helper function and put this as a docstring.

File: indico/modules/rb/controllers/user/reservations.py [hunk 4]
```diff
@@ -457,9 +457,30 @@ def _checkParams(self):
...
+        reservation_duration = self._reservation.end_dt - self._reservation.start_dt
+        date_delta = date.today() - self._reservation.start_dt.date()
+        start_dt = self._reservation.start_dt + date_delta
+        end_dt = start_dt + reservation_duration
+        if start_dt < datetime.now():
+            new_start_dt = round_up_to_minutes(datetime.now(), 15) + timedelta(minutes=15)
+            time_delta = new_start_dt - start_dt
+            start_dt = new_start_dt
+            end_dt = end_dt + time_delta
+        return start_dt, end_dt
```
Comment (ThiefMaster) [inline indico/modules/rb/controllers/user/reservations.py]: If you return a dict here you can simply use `changes.update(self._update_datetime())`

File: indico/modules/rb/controllers/user/reservations.py [hunk 5]
```diff
@@ -457,9 +457,30 @@ def _checkParams(self):
...
         return RHRoomBookingNewBookingSimple._get_view(self, clone_booking=self._reservation, **kwargs)
 
+    def _update_datetime(self):
+        """Make necessary changes to reservation's start and end datetime.
+
+        Move the start date to the current day, adjust the end date
+        accordingly and if the start datetime is still in the past, change the
+        start and end time to the current (rounded up) time.
+
+        :return: A tuple with the new start_dt, end_dt
```
Comment (ThiefMaster) [inline indico/modules/rb/controllers/user/reservations.py]: "a tuple with the new times" (or "a dict" if you change it)

Comment (nop33) [inline indico/modules/rb/controllers/user/reservations.py]: "times", rly? One time, two times? :stuck_out_tongue: 
Anyway, it's datetime, the date is also altered.

Comment (ThiefMaster) [inline indico/modules/rb/controllers/user/reservations.py]: We recently said that "time" includes dates too :stuck_out_tongue: 

but "with the new datetimes" is fine too

Comment (nop33) [inline indico/modules/rb/controllers/user/reservations.py]: In this context, time refers to the time on the clock and date to the date of the calendar.

Comment (nop33) [inline indico/modules/rb/controllers/user/reservations.py]: Plus you cannot say "the times/datetimes of the booking"

File: indico/modules/rb/controllers/user/reservations.py [hunk 6]
```diff
@@ -457,9 +457,30 @@ def _checkParams(self):
     def _get_view(self, **kwargs):
         return RHRoomBookingNewBookingSimple._get_view(self, clone_booking=self._reservation, **kwargs)
 
+    def _update_datetime(self):
```
Comment (OmeGak) [inline indico/modules/rb/controllers/user/reservations.py]: What happens if the room is booked for the updated datetime?

Comment (ThiefMaster) [inline indico/modules/rb/controllers/user/reservations.py]: out of the scope; you get a red bar and adapt the dates accordingly

Comment (OmeGak) [inline indico/modules/rb/controllers/user/reservations.py]: :+1:

**Comment by ThiefMaster** [comment]
Merged b93e8e08f90305c915d4c6e17d802d1b22fe0c40

**Comment by nop33** [comment]
When this get's into prod we should remember to close `INC0962765`

**Comment by ThiefMaster** [comment]
It is in prod

**Comment by nop33** [comment]
That was.. worth mentioning :stuck_out_tongue:

**Pull request closed without merging**